Implement abstraction of transactions by adding tx interface

begmaroman opened the pull request:
# Description

Geth introduced concept of envelope+payload for transactions, before EIP-1559 so that interface of transaction remains the same for the "outer" world and created common interface for both legacy and dynamic fee (EIP-1559) transactions:

- https://github.com/ethereum/go-ethereum/blob/master/core/types/transaction.go (this is envelope struct I was referring to and it embeds either legacy tx, access list tx or dynamic fee/EIP-1559 tx)

These implementations are embedded as the payload of the transaction object itself:

- EIP-1559 transactions https://github.com/ethereum/go-ethereum/blob/master/core/types/tx_dynamic_fee.go
- legacy transactions(gas price only transactions, pre-EIP-1559 hardfork) https://github.com/ethereum/go-ethereum/blob/master/core/types/tx_legacy.go

Introduce `Transaction` interface and add two implementations: state tx and legacy tx 

# Changes include

- [ ] Bugfix (non-breaking change that solves an issue)
- [ ] Hotfix (change that solves an urgent issue, and requires immediate attention)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Checklist

- [x] I have assigned this PR to myself
- [x] I have added at least 1 reviewer
- [x] I have added the relevant labels
- [ ] I have updated the official documentation
- [ ] I have added sufficient documentation in code

## Testing

- [x] I have tested this code with the official test suite
- [x] I have tested this code manually

Nemanja0x committed:
Move PolyBFT to Edge (#774)

* Move PolyBFT to edge

Co-authored-by: Igor Crevar <crewce@gmail.com>
Co-authored-by: Sasa Prsic <sasa@ethernal.tech>
Co-authored-by: Stefan Negovanović <stefan@ethernal.tech>
Co-authored-by: Victor Castell <victor@victorcastell.com>

igorcrevar committed:
Build block with Txs in PolyBFT (#789)


Co-authored-by: Stefan Negovanović <stefan@ethernal.tech>
Co-authored-by: Sasa Prsic <sasa@ethernal.tech>

begmaroman committed:
EVM-48: Add RootChain helper (#773)

igorcrevar committed:
EVM-57 PolyBft state transactions introduction (#809)

* EVM-57 PolyBFT bridge enabled state txs

Co-authored-by: Sasa Prsic <sasa@ethernal.tech>

* EVM-57 PolyBFT bridge enabled state txs

Co-authored-by: Sasa Prsic <sasa@ethernal.tech>

* tidy.

* pr fixes

* pr fix 2

* small fix

* TransactionType should be populated in receipt

* nonce fix

* write change original tx fix

* createStateTransactionWithData change

* pr fix more

* revert ComputeHash for tx because id does not really matter

Co-authored-by: Sasa Prsic <sasa@ethernal.tech>

Commit:
Migrate polybft e2e framework to edge (#797)

Back-port the e2e framework + tests from v3 to edge as a separate set of tests.

Stefan-Ethernal committed:
v3 smart contracts integration (#796)

* Initial changes

* Init v3-contract submodule

* Uptime changes and linter fixes

* Use local build and test workflows (temporary)

* Test workflow expand with init git submodules

* Native transfer precompile

* CommitTransaction in blockBuilder

* adding workflow call secrets input (#799)

* Add BLS aggregated signatures verification precompile

* Change required gas for introduced precompiles

* Add console precompile

* Additional functions in bls package

* Register validator command

* Linter fixes, register subcommand for register validator command

* Adapt genesis command

* Cleanup

* Change addresses of sidechain contracts so that they don't clash with precompile addresses

* Marshal from address in transaction

* Remove CommitTransaction and use WriteTx instead

* Fix panic in register validator command

* Get rid of panics in register validator

* Increase gas limit for ChildValidatorSet initialization

* Cleanup GenesisPostHookFactory

* Fix register validator command

* Change gas limit to 100M for initialize child validator set

* Change make compile v3-contracts

* Recover CI (1st attempt)

* Recover CI (2nd attempt)

* Recover CI (3rd attempt)

* Cleanup Makefile (remove compilation of SCs when unneeded). Remove cloning submodules in lint workflow

* Recover lint

* Remove init submodules from test workflow

* Address some go security issues

* small make fix

* init root contracts command

* Remove rootchain contracts deployment from rootchain server command

* Provide new url for smart contracts

* Introduce constants for register validator and contracts root folder

* Comments

* Comments (part 2)

* Reference dev-ethernal commit from core-contracts

* Fix polybft e2e workflow

* Fix genesis cmd invocation from e2e framework

* contract_initializer fix

* Fix polybft e2e tests

* Comments (part 3)

* Populate validators balance based on provided genesis configuration.
Comments (part 4)

* Fix balance unmarshal for polybft.Validator

* Initialize native transfer contract

Co-authored-by: Goran Rojovic <goran.rojovic@ethernal.tech>
Co-authored-by: Epikichi <78609649+epikichi@users.noreply.github.com>
Co-authored-by: Igor Crevar <crewce@gmail.com>

Stefan-Ethernal committed:
EVM-129 Implement bridge workflow (#828)

* Add bridge flag.

* Populate bridge config.

* Fix bridge address.

* rootchain panic fixes

* changes. deploying contracts to root still does not work in test

* fix deploying contracts and running bridge. Emit command change

* Read balance from genesis

* Remove Geth reference from Edge module

* bridge test fixes

* remove test-bootnode. add WithBootnodeCount to test-cluster. Change network_test to be more relaxed

Co-authored-by: Sasa Prsic <sasa@ethernal.tech>
Co-authored-by: Igor Crevar <crewce@gmail.com>

Stefan-Ethernal committed:
Transfer unit tests from v3 to Edge (#825)

* test-bridge fix + temporary fix

* Read balance from genesis

* state, state transaction, event tracker

* FSM tests (part 1)

* ConsensusRuntime tests

* FSM tests (part 2)

* FSM tests (part 3)

* SystemState tests

* ValidatorAccount and ValidatorSnapshot tests

* buildLogsFromReceipts test

* FSM tests (part 4)

* FSM fix remaining tests

* t.parallel

* Introduce parallel execution in tests (part 2)

* Lint fixes

* Rename test helpers file to exclude it from build

* Add test suffix to mocks

* Install solidity compiler on test workflow

* Assign execute permission to setup-ci.sh script

* polybft tests

* all tests merged

* setting epoch snapshot added

* Lint fixes

* Use constant for IstanbulExtraVanity

* Description of polybft test

Co-authored-by: Igor Crevar <crewce@gmail.com>
Co-authored-by: Nemanja0x <nemanja@ethernal.tech>
Co-authored-by: Goran Rojovic <goran.rojovic@ethernal.tech>

goran-ethernal committed:
[Checkpointing] `ExitEvent` storage and proof generation (#838)

* Exit event storage and query functions

* ExitProof generation

* Reorganize tests

* Optimize key composition

* Extract common function for decoding state sync and exit events

* Comments fix

* Comments fix

* [Checkpointing] JSON RPC point for generating exit proofs (#844)

* JSON RPC point for generating exit proofs

* Comments fix

Co-authored-by: Stefan Negovanović <stefan@ethernal.tech>

Stefan-Ethernal committed:
Checkpoint data signing (#840)

* Add CheckpointData to Extra

* Don't store chainID, blockNumber and blockHash to the blockchain

* Set checkpoint data to the Header.Extra

* Fix and simplify FSM unit tests

* getSignHash static function, fix verify header implementation

* Increase block waiting time in bridge test

* Remove unnecessary condition

* Set checkpoint to each extra

* Sign checkpoint hash instead of aggregated hash

* Make nextValidatorsHash same as currentValidatorsHash for non-epoch ending blocks

* Set event root to the checkpoint object

goran-ethernal committed:
[Checkpoints] Parse receipt logs and insert new exit events (#852)

* Parse receipt logs and insert exit events

* Comments fix

* Comments fix

* Comments fix

* Remove TODO

* Comments fix

* Comments fix

* Comments fix

Stefan-Ethernal committed:
Consolidate BLS verification ABI type (#867)

Stefan-Ethernal committed:
Checkpoint submission (#861)

* Change error messages for checkpoint hash calculation

* Determine checkpoint blocks

* Checkpoint submission (draft)

* Introduce checkpoint manager

* Lint fixes

* Fix tests

* Checkpoint manager checkpoint data ABI encode unit tests

* Unit test for getCurrentCheckpointID

* Unit test for submitCheckpoint

* Remove leftover code

* Comments

* Determine epoch ending blocks by comparing epoch numbers in adjacent headers

* Micro optimization

* Skip querying first block for end of epoch

* Optimization

* Simplify checkpoints offset

* Adapt comment

* Checkpoint block detection optimization (#872)

* Simplify checkpoint block detection

* Resolve conflicts and pass tests

Co-authored-by: Boris <boris@Boriss-MacBook-Pro.local>

* Simplify

* Submit pending checkpoints in unit test

* Extract parentHeader retrieval outside of the for

* Submit end of epoch blocks for pending checkpoints

* Update submitCheckpoint comment

Co-authored-by: Victor Castell <victor@polygon.technology>

* Feedback

* Separate import declarations

* Create var block for `parentExtra` and `parentHeader`

Co-authored-by: Victor Castell <victor@polygon.technology>

Co-authored-by: Goran Rojovic <goran.rojovic@ethernal.tech>
Co-authored-by: Boris <boris@Boriss-MacBook-Pro.local>
Co-authored-by: Victor Castell <victor@polygon.technology>

Stefan-Ethernal committed:
Merge branch 'develop' into feature/v3-parity

Stefan-Ethernal committed:
Mock Host implementation in native transfer test (#887)

* Mock Host implementation in native transfer test

* Lint fix

* Fix polybft e2e tests

Commit:
Implement the logic for quorum check (#837)

* Implement the logic for quorum check

Depending on the message type it will implement different kinds of quorum checks.

* Remove MaximumFaulty

Stefan-Ethernal committed:
Integrate go-ibft consensus engine to polybft consensus protocol (#891)

* EVM 149 - syncer go ibft coordination (#843)

* consensus runtime backend metohod (not buildable)

* transport

* polybft - run cycle

* transport fix 2

* small changes

* less complex ibft consennsus wrapper.Example: https://replit.com/@crewce/Newest#main.go

* small changes

* runSequence better

* remove pbftTransportWrapper

* noone likes temp variables

* Comment fixed

* jm fixes

Co-authored-by: Igor Crevar <crewce@gmail.com>

* EVM-151 Use (or not) edge hooks support (#849)

* FSM that supports go-ibft backend interface (#865)

* consensus runtime backend metohod (not buildable)

* transport

* polybft - run cycle

* transport fix 2

* less complex ibft consennsus wrapper.Example: https://replit.com/@crewce/Newest#main.go

* Implement MessageConstructor interface.

* Tidy linter errors.

* runSequence better

* remove pbftTransportWrapper

* Add quorum size implementation.

* Remove one part of pbft.NodeID reference.

* Add some comments.

* IsValid, Build proposal on FSM

* Add block validation.

* Add stub for IsValidSender method.

* Sender validation is always true.

* fsm - Insert block

* simplify build proposal

* Tidy logs.

* IsValidSender

* pbft.Proposal removed

* Tidy errors.

* restartEpoch if epoch is nil, remove hook in fsm

* initialization fix

* Fix compare.

* quorum fixes and error propagation

* evm-151 changes + additional needed in this task

* Increase test timeout.

* bring back some part of fsm

* just execute syn test 100 times in a row

* Tidy part of errors.

* Revert e2e test.

* Remove unused code and move message creator interface implementation.

* minor fix

* Add comments.

* code organization - valdateSender + recoverAddress

* Fix compile after merge.

* small change

* restart epoch on better place. lastBuiltBlock and epoch should not be nil

* small reorg in polybft

* pr fix

* Update comments

* pr fixes no 2

* Add voting power to the validators (#880)

* Add voting power to the validator account

* Rename ValidatorAccount to ValidatorMetadata

* Set voting power to the existing test

* Comment fixes

* Rename files

* Address comments

* Remove extra space from the comment

* pr fixes

* Rename GetValidatorAccount to GetValidatorMetadata

Co-authored-by: Nemanja0x <nemanja@ethernal.tech>
Co-authored-by: Igor Crevar <crewce@gmail.com>
Co-authored-by: Stefan Negovanović <stefan@ethernal.tech>
Co-authored-by: Stefan Negovanović <93934272+Stefan-Ethernal@users.noreply.github.com>

* Accommodate go-ibft `BuildProposal` updated function (#893)

* Populate round info and change proposal hash

* Adopt currentRound from BuildProposal

* Fix logs

* Fix proposal hash calculation inside IsValidProposalHash

* Update go-ibft version

* Igor's feedback addressed

* Merge fix

* EVM-187 Handling bundles on both sync and consensus (#882)

* EVM-187 Handling bundles on both sync and consensus

* consensus_runtime tests

* UT fix

Co-authored-by: Goran Rojovic <goran.rojovic@ethernal.tech>

* Implement HasQuorum function in consensusRuntime

Co-authored-by: Nemanja Nedic <93836814+Nemanja0x@users.noreply.github.com>
Co-authored-by: Igor Crevar <crewce@gmail.com>
Co-authored-by: Saša Pršić <93726535+0xSasaPrsic@users.noreply.github.com>
Co-authored-by: Nemanja0x <nemanja@ethernal.tech>
Co-authored-by: Goran Rojovic <goran.rojovic@ethernal.tech>

Stefan-Ethernal committed:
Merge branch 'develop' into feature/v3-parity

Stefan-Ethernal committed:
Merge branch 'develop' into feature/v3-parity

Stefan-Ethernal committed:
Remove Polybft dummy smart contracts (#909)

Stefan-Ethernal committed:
Change ValidatorMetadata methods to pointer receivers (#911)

igorcrevar committed:
EVM-152 Fix Unit Tests (#910)

Co-authored-by: Goran Rojovic <goran.rojovic@ethernal.tech>

epikichi committed:
adding devnet-v3 workflow for parity testing. Adding needed secrets for other workflows (#916)

Commit:
Introduce local development cluster script (#912)

* Introduce local development cluster script

Stefan-Ethernal committed:
Fix assigned boot nodes port number in genesis (#930)

Commit:
Local cluster - adapt port for ibft to polybft (#931)

Stefan-Ethernal committed:
Merge branch 'develop' into feature/v3-parity

Stefan-Ethernal committed:
Merge branch 'develop' into feature/v3-parity

Stefan-Ethernal committed:
Fix unit test for event tracker (#935)

* Fix linters

* Fix TestEventTracker_TrackSyncEvents

* Comment

igorcrevar committed:
EVM-232 Keep polybft bls and ecdsa keys in separate secrets manager namespaces (#953)

Stefan-Ethernal committed:
Track updates for validators (#939)

* Implement equality checks for PublicKey and ValidatorMetadata.
Expand ValidatorSetDelta

* Handle updates in ApplyDelta

* Additional tests

epikichi committed:
Epikichi/edge 973 change v3 parity workflow to devnet (#957)

* reflecting changes to use ephemeral environment instead

* Epikichi/edge 964 implement multiple erc mode support (#954)

* adding support for ERC20 and ERC721 mode for Pandora's Box stress testing

* adding in default for devnet ephemeral anticipation

* fixing markdown format

* reverting for a separate PR

* adding changes for devnet ephemeral workflow to allow dispatching

* adding correct job names, removed dependencies since concurrency protection is set on Pandora' workflow env-wide

* reflecting chnages from develop

igorcrevar committed:
EVM-220 TestClusterBlockSync/BLS fails in voting power branch (#926)

0xSasaPrsic committed:
Fix data race to polybft unit test (#965)

* Fix some race conditions.

Stefan-Ethernal committed:
Polybft voting power (#944)

voting power feature

Stefan-Ethernal committed:
Align `v3-parity` with `develop` branch (#971)

* Epikichi/edge 964 implement multiple erc mode support (#954)

* adding support for ERC20 and ERC721 mode for Pandora's Box stress testing

* adding in default for devnet ephemeral anticipation

* fixing markdown format

* reverting for a separate PR

* introducing workflow dispatch trigger for Devnet V3 workflow. (#927)

* introducing workflow dispatch trigger for Devnet V3 workflow.

* adding ephemeral environment naming

* adding pandora modes

* Add performance tests to devnet deployments (#918)

* added performance tests to devnet deployment
* removing depricated loadbot performance tests

* Docker setup: Wait for genesis file to exist before starting nodes (#949)

* Docker setup: Wait for genesis file to exist before starting nodes

* Docker setup: don't run init if genesis file already exists

* adding conditional dependency for compose to wait for successful  node

Co-authored-by: epikichi <ji@epik.dev>

Co-authored-by: Epikichi <78609649+epikichi@users.noreply.github.com>
Co-authored-by: ZeljkoBenovic <47507986+ZeljkoBenovic@users.noreply.github.com>
Co-authored-by: mediremi-antithesis <107197847+mediremi-antithesis@users.noreply.github.com>
Co-authored-by: epikichi <ji@epik.dev>

Stefan-Ethernal committed:
Merge branch 'develop' into feature/v3-parity

Stefan-Ethernal committed:
Fix e2e tests

Stefan-Ethernal committed:
Set address to the result instance on applyCreate

Commit:
Remove State Sync execution step (#964)

* Remove commitment execution related stuff
* Add the RPC endpoint to obtain StateSync Proofs
* Rootchain data to gitignore
* Return the state sync along with proof
* Better params for cluster script
* E2E test executing state syncs
* Fix tests that rely on bundle execution
* Remove commitment execution related stuff

igorcrevar committed:
Replace BLS library with github.com/kilic/bn254 (#978)

goran-ethernal committed:
Integrate with CheckpointManager.sol (#952)

* Integrate with CheckpointManager.sol

* Hex encode ECDSA private key

* Skip failing tests for checkpoint manager

* Fix submit function signature

* Remove logs and set From field to submit checkpoint transaction

* Pass signer key to rootchain.SendTxn function

* Pull console precompile Geth docker image

* Populate voting power on checkpoint manager initialization

* Hash change

* Minor fixes (ordering of abi ancoded parameters)

* Populate voting power on genesis block

* Fix balance scaling

* Fix TestCheckpointManager_isCheckpointBlock

* Remove sorting by addresses on validators snapshot calculation

* Sort validators by addresses on genesis header population

* Sort validators by addresses on rootchain initialization

* Logging...

* Update core-contracts

* Fix checkpoint manager unit tests

* Change comment + unit tests for helper functions

* Revert usage of Geth console fork

* Include Geth console fork again

* Rename latestCheckpointBlockNumber to currentCheckpointBlockNumber

* Remove logs and revert to use official Geth image for rootchain server

* Code cleanup

* Add meaningful logs to the checkpoint manager

* not buildable

* full functional bn254

* Use proper domain when initializing CheckpointManager contract

* Fix go mod

* Revert helper bash scripts

* Fix e2e tests

* Address Igor's comment

* Provide `epochSize` to the `ChildValidatorSet` (#983)

* Provide epochSize to the ChildValidatorSet

* Rename newBls map key

* Rename parameters

* Point to our branch in core-contracts

Co-authored-by: Goran Rojovic <goran.rojovic@ethernal.tech>

Co-authored-by: Stefan Negovanović <stefan@ethernal.tech>
Co-authored-by: Nemanja0x <nemanja@ethernal.tech>
Co-authored-by: Igor Crevar <crewce@gmail.com>
Co-authored-by: Stefan Negovanović <93934272+Stefan-Ethernal@users.noreply.github.com>

goran-ethernal committed:
Merge remote-tracking branch 'origin/develop' into feature/v3-parity

Stefan-Ethernal committed:
Change genesis premine logic (#981)

0xSasaPrsic committed:
Fix data race in fsm (#982)

* Fix data race in several places accessing fsm and fix copy validators function.

* tidy linter.

* Fix tests after changes and user RLock.

Commit:
Removing `GenesisTarget` and adding the `NodeID` field to `Validator` struct (#985)

* Refactored GenesisTarget removing it and adding the NodeID field to Validator struct.
* Refactored generatePolyBFTConfig to use the new struct and simplify the code and take into account validators passed from params
* Add NodeID as a field needed for validators passed from params
* Tune timeout for bridge test

Co-authored-by: Stefan Negovanović <stefan@ethernal.tech>

Stefan-Ethernal committed:
Provide rootchain IP address from CLI (#989)

* Refactor RootchainInteractor abstraction

* Lint fix

* Provide JSON RPC address flags

* Remove lock from rootchain interactor.
Send transaction remove nonce parameter.

* Address comments

* Reduce amount of affected files

* Implement TxRelayer

* Address comments (part 2)

* Create txn with nonce set

* Option pattern in TxRelayer

* Use txRelayer in e2e tests

* Let TxRelayer take care of setting the nonce

* Fix tx pool e2e tests

begmaroman committed:
Added london hardfork to the chain config (#999)

Stefan-Ethernal committed:
Fix pending checkpoints submission (#992)

* Fix epoch ending blocks detection in pending checkpoints

* Update TestCheckpointManager_submitCheckpoint

* Remove redundant signerAddress

* Make unit test more understandable

begmaroman committed:
Added transaction abstraction interface

begmaroman committed:
Fixed all compilation errors

begmaroman committed:
Fixed linter

begmaroman committed:
Fixed linter

begmaroman committed:
Fixed linter

begmaroman committed:
Fixed linter

begmaroman committed:
Fixed test

begmaroman committed:
Fixed test

begmaroman committed:
Fixed test

begmaroman committed:
Fixed test

begmaroman committed:
Fixed test

begmaroman committed:
Fixed test

begmaroman committed:
Fixed linter

begmaroman committed:
Fixed linter

begmaroman committed:
Fixed linter

begmaroman committed:
Fixed linter

begmaroman committed:
Fixed linter

begmaroman committed:
Fixed linter

begmaroman committed:
Fixed linter

begmaroman committed:
Minor update

begmaroman committed:
Minor update

begmaroman committed:
Minor update

begmaroman committed:
Minor update

begmaroman committed:
Minor fixes

begmaroman committed:
Minor fixes

begmaroman committed:
Minor fixes

Stefan-Ethernal committed:
Update core-contracts (#1007)

begmaroman committed:
Fixed linter

begmaroman committed:
Fixed linter

begmaroman committed:
Fixed linter

begmaroman committed:
Fixed test

JekaMas commented on types/transaction.go:
```diff
@@ -12,73 +13,102 @@ import (
+}
+
+// NewTx creates a new transaction.
+func NewTx(inner TxData) *Transaction {
```
when we start using go 1.19 it'd be nice to use generics here and in NewTxWithSender. Maybe for Transaction also and have something like

```
type Transaction[T TxData] struct {
...
    inner T
...
}

// maybe even the interface
type txData interface {
        txType() TxType // returns the type ID
    	input() []byte
	gas() uint64
	setGas(gas uint64)
	gasPrice() *big.Int
	gasTipCap() *big.Int
	gasFeeCap() *big.Int
	value() *big.Int
	nonce() uint64
	to() *Address

	rawSignatureValues() (v, r, s *big.Int)
	setSignatureValues(v, r, s *big.Int)
}

type TxData[T txData] interface {
	Copy() T   // creates a deep copy and initializes all fields
        txData
}
```

Please, make a task for it.

begmaroman replied:
Very nice point! Done: EVM-301

goran-ethernal committed:
Bundle build fix (#1014)

goran-ethernal committed:
Check for nil proposal (#1009)

ferranbt committed:
Remove gasLimit and gas fields from txrelayer call (#1020)

ferranbt committed:
Add default gas and gasPrice to local transactions (#1021)

ferranbt committed:
Use geth console fork as default with flag to disable (#1022)

Stefan-Ethernal committed:
Checkpoint submission e2e test (#990)

* Checkpoint submission e2e test

* Utilize TxRelayer in test

* Fix init contracts command

* Fix e2e test

Commit:
Merge branch 'develop' into feature/v3-parity

Stefan-Ethernal commented on types/transaction.go:
```diff
@@ -150,9 +161,94 @@ func (t *Transaction) Size() uint64 {
+}
+
+// setDecoded sets the inner transaction and size after decoding.
+func (t *Transaction) setDecoded(inner TxData, size uint64) {
```
What's the purpose of `size` param here? As I can see we are always providing 0 for it.

begmaroman replied:
Makes sense, removed

Stefan-Ethernal commented on types/rlp_unmarshal.go:
```diff
@@ -307,79 +307,156 @@ func (t *Transaction) UnmarshalRLPFrom(p *fastrlp.Parser, v *fastrlp.Value) erro
+			return err
+		}
+
+		// to
+		if vv, _ := v.Get(3).Bytes(); len(vv) == 20 {
+			// address
+			addr := BytesToAddress(vv)
+			lTx.To = &addr
+		} else {
+			// reset To
+			lTx.To = nil
+		}
```
Nitpick: does it make sense to bring `to` field unmarshal above `value` and `input` fields, because it is on index 3, comparing `value` and `input` which are on indices 4 and 5 respective?

begmaroman replied:
Done

Stefan-Ethernal commented on types/transaction.go:
```diff
@@ -87,48 +117,29 @@ func (t *Transaction) ComputeHash() *Transaction {
+	return &Transaction{
+		inner: t.inner.Copy(),
+		time:  t.time,
+		hash:  t.hash,
```
Create copy of `hash` field as well?

begmaroman replied:
We copy all data here so the hash would remain the same

Stefan-Ethernal replied:
I was referring to mutation of `hash` field on the `Transaction` instance itself and in that regard copying that byte slice, but that's probably not a valid use case...

begmaroman replied:
It is an array so this statement is not applicable. It copies an original array.

Stefan-Ethernal replied:
I was having this example on mind: https://go.dev/play/p/JP6kGT2XVwx. Am I missing out something?

begmaroman replied:
Yes, there is a slice in your example, but the hash is an array, they behave differently.

Stefan-Ethernal replied:
Ok I see now: https://go.dev/play/p/UlXTOwSOih0 🙂

Stefan-Ethernal commented on types/rlp_encoding_test.go:
```diff
@@ -59,8 +59,10 @@ func TestRLPMarshall_And_Unmarshall_Transaction(t *testing.T) {
 
-	txn.Hash = unmarshalledTxn.Hash
-	if !reflect.DeepEqual(txn, unmarshalledTxn) {
+	// txn.Hash = unmarshalledTxn.Hash
```
Think that you could have left this line in this particular case. However if `ComputeHash` is going to be kept instead, then just remove this commented line.

begmaroman replied:
ok, done

Stefan-Ethernal commented on txpool/event_subscription_test.go:
```diff
@@ -141,13 +143,14 @@ func TestEventSubscription_ProcessedEvents(t *testing.T) {
-				t.Fatalf("Unable to wait for events to be processed, %v", err)
-			}
+			})
+			require.NoError(t, err)
```
Would you like to keep this error message as third parameter of `require.NoError`?

`Unable to wait for events to be processed`

begmaroman replied:
Done

Stefan-Ethernal commented on types/transaction.go:
```diff
@@ -12,73 +13,102 @@ import (
 
 func (t *Transaction) IsLegacyTx() bool {
-	return t.Type == LegacyTx
+	return t.inner.txType() == LegacyTxType
 }
 
 func (t *Transaction) IsStateTx() bool {
-	return t.Type == StateTx
+	return t.inner.txType() == StateTxType
 }
```
Just thinking out loud and I'm not biased towards any approach:
maybe remove these functions and rely on `Type()` function instead whenever needed? Otherwise, soon we are about to add `DynamicFeeTxType` and possibly add `IsDynamicFeeTx` helper function...

begmaroman replied:
> Otherwise, soon we are about to add DynamicFeeTxType and possibly add IsDynamicFeeTx helper function...

That's right. Why not?

Stefan-Ethernal replied:
I see it as a redundant in a way, since we can always query `TxType`, but however you want.

begmaroman replied:
Well, this is what it is right now. I would change it within one of the next PRs anyways.

Stefan-Ethernal commented on command/e2e/register_validator.go:
```diff
@@ -294,22 +294,22 @@ type txnSender struct {
-func (t *txnSender) sendTransaction(txn *types.Transaction) asyncTxn {
-	if txn.GasPrice == nil {
-		txn.GasPrice = big.NewInt(defaultGasPrice)
+func (t *txnSender) sendTransaction(txnData *types.LegacyTx) asyncTxn {
```
Didn't want to have `*types.Transaction` here as a parameter?

begmaroman replied:
Not really because it is used within this package only and adds some extra fields population within the function itself. It would not be possible if we pass the common transaction model.

Stefan-Ethernal reviewed (commented):
Generally looks good, although leaving a couple of comments to consider.

JekaMas committed:
git modules: shallow and fix branch (#987)

* shallow and fix branch

* feat-polybft-release

Stefan-Ethernal committed:
Remove bool return value from deliverMessage signature (#1027)

begmaroman committed:
Addressing comments

begmaroman committed:
Fixed nit

0xSasaPrsic committed:
Fix/immutable trie race (#1017)

* Add lock to trie state.

Stefan-Ethernal committed:
Use total stake for voting power (#1038)

* Use total stake as voting power

* Lower log level to debug log in local cluster script

* Fix unit test

Stefan-Ethernal committed:
RLP Marshal/Unmarshal update field in `ValidatorSetDelta` (#1026)

* RLP Marshal/Unmarshal update field in ValidatorSetDelta

* Comment + additional unit test

Stefan-Ethernal committed:
Remove unused fields (#1039)

begmaroman committed:
Introduce metrics for BoltDB for the state (#1045)

Stefan-Ethernal committed:
Merge branch 'develop' into feature/v3-parity

Stefan-Ethernal committed:
Expose `PolyBFT` metrics to Prometheus API (#1040)

* Introduce metrics to polybft

* Add comment

* Update comment

* Remove total state syncs metrics

* Include metrics prefix and rename file

* Address comment

* Fix consensus runtime tests

goran-ethernal committed:
EVM-308 Add commands we are missing (#1050)

* Added new commands

* Merge params and result files

* Reorganize commands

* Comments fix

Stefan-Ethernal committed:
Fix loop closure violations in unit tests (#1053)

* Fix loop closures

* Make TestExtra_CreateValidatorSetDelta_Cases deterministic

* Avoid for loop when assigning voting power

* Minimize changes

* Fix

b00ris committed:
Integrate L2 exit workflow (#1024)

* e2e exit test

* fix lint

* add static L1 exit contract data

* add contract comilation to e2e tests

* fix ci

* save artifacts in client

* fix tests

* review fix

* review fix

* Update consensus/polybft/contractsapi/gen/main.go

Co-authored-by: Stefan Negovanović <93934272+Stefan-Ethernal@users.noreply.github.com>

* review fix

Co-authored-by: Stefan Negovanović <93934272+Stefan-Ethernal@users.noreply.github.com>

Stefan-Ethernal committed:
Add pending checkpoints assertions to checkpoint submission e2e test (#1052)

Stefan-Ethernal committed:
Decouple rootchain contracts deployment from development environment (#1005)

* Rootchain init contracts read validators from the genesis configuration.
Introduce manifest.json.

* Introduce admin-key flag to the rootchain init-contracts

* Revert polybft params changes

* Fix comment

* Populate rootchain admin address to manifest

* Update Manifest struct and adapt init-contracts workflow

* Remove sidechain smart contract addresses from genesis spec

* Update genesis command and cleanup genesis from unnecessary parameters

* Remove rootchain contracts hardcoded addresses

* Decouple emit command from contant StateSender address

* Manifest command

* Polybft manifest command

* Adapt e2e tests workflow

* Update comment

* Adapt cluster script and change flags descriptions in manifest command

* Comments

* Adapt TestE2E_CheckpointSubmission

* Remove manifest.json in cluster script

* Initial version of polybft readme

* Update readme with a precondition step

* Address comments (part 1)

* Update manifest command short help

* Comments

* Use sidechain addresses from the polybft config in fsm UTs

* Minimize changes

* Merge fix

Co-authored-by: Goran Rojovic <goran.rojovic@ethernal.tech>

igorcrevar committed:
EVM-302  Save proposer priorities after each block (#1018)


* EVM-302 Save proposer priorities after each block

Co-authored-by: Nemanja0x <nemanja@ethernal.tech>
Co-authored-by: Stefan Negovanović <stefan@ethernal.tech>
Co-authored-by: Sasa Prsic <sasa@ethernal.tech>

Stefan-Ethernal committed:
Remove GossipMessage (#1057)

Stefan-Ethernal committed:
Consolidate `ExtendByteSlice` implementations (#1058)

* Consolidate ExtendByteSlice implementations

* Add comment and fix

* Introduce trim parameter (fix unit tests)

* Update bitmap unit tests

* Address comment

* Add a unit test

Stefan-Ethernal committed:
Remove readContractBytecode (#1061)

Nemanja0x committed:
Evm 303: Fix race conditions after a block is inserted (#1047)


Co-authored-by: Igor Crevar <crewce@gmail.com>
Co-authored-by: Stefan Negovanović <stefan@ethernal.tech>
Co-authored-by: Sasa Prsic <sasa@ethernal.tech>

0xSasaPrsic committed:
Feature/validator set changes (#1048)

* Update priorities when validator set changes.

Commit:
Remove code related to bundles for state syncs (#1013)

Related to RFC-139, we're removing bundles concept from the codebase.
We're introducing the latest Smart Contracts version in this PR

ferranbt committed:
Create a full block object during polybft (#1063)

* Create a full block object during polybft

* Remove print

* Remove now

* Fix tests

Stefan-Ethernal committed:
Fix segmentation fault panic on `ValidateSender` (#1054)

* Add nil check for FSM

* Different approach: consensus runtime validates message sender.
Remove Includes function from validatorSet.

* Subscribe IBFT topic only after epoch is initialized

* Renames and comment fixes

* Revert sender validation to the fsm

* Rebase fix

* Add log

goran-ethernal committed:
RFC173 Checkpoint consensus issue (#1064)

* RFC 173 changes

* Comments fix

* Comments fix

Stefan-Ethernal committed:
Fix uptime total blocks calculation (#1069)

* Fix uptime total blocks calculation

* Update test for uptime calculation

* Simplify

* Simplify unit test

* Typo fix

Commit:
Implement the relayer (#1044)

* Implement the relayer
* The statesyncrelayer package will reside inside polybft

0xSasaPrsic committed:
Immutable trie simplification (#1068)

* Refactor immutable trie.

* Remove state from trie.

* Tidy.

* Remove storage from trie.

Stefan-Ethernal committed:
Remove voting power scaling (#1060)

* Change voting power from uint64 to big.Int

* Provide desired scaling factor to validator set

* Fix unit tests after rebase

* Fix validatorSetToABISlice comment

* Introduce relative voting power

* Change ProposerPriority to big.Int

* Fix log

* Fix computeMaxMinPriorityDiff

* Remove max total voting power hard limit

* Use truncated division for rescaling priorities

* Add ceil to quorum size calculation

* Revert unnecessary changes

* Remove safe_math

* Compress quorum calculation

Co-authored-by: Victor Castell <victor@polygon.technology>

* Extract helper function for big int with ceil division

Co-authored-by: Victor Castell <victor@polygon.technology>

epikichi committed:
Epikichi/edge 988 add polybft consensus support to edge (#1051)

* Added support for polybft consensus docker cluster
* added customisation options
* added/edited readme

Co-authored-by: ZeljkoBenovic <sakal87@gmail.com>
Co-authored-by: ZeljkoBenovic <47507986+ZeljkoBenovic@users.noreply.github.com>

Stefan-Ethernal committed:
Implement `TestE2E_Bridge_ChangeVotingPower` (#1075)

* Implement TestE2E_Bridge_ChangeVotingPower

* Increase epoch rewards

* Add log to the fsm.getCurrentValidators

* Move default epoch reward to genesis cmd

* Dump validators to log only when debug logs enabled

* Multiple validators are staking

Stefan-Ethernal committed:
Remove unnecessary panics from polybft (#1077)

* Register validator remove panics and expose json rpc flag

* Remove panic from blockchainWrapper

* Remove json rpc flag from register validator cmd

* Polybft log instead of panic if block synchronization fails

* Remove unused vars

* Handle createValidatorSetDelta error

* Rename test

* Revert panic in stateProvider

goran-ethernal committed:
Dynamic epoch size (#897)

* Decouple uptime calculation from fixed epoch size

* Tests fix

* Rename isEndOfEpoch

* Return error on apply delta to validator set

* Comments fix

* Comments fix

* Comments fix

* Comments fix

* Comments fix

* Bug fix

* Comments fix

* Rebase fix

* Rebase fix

* Rebase fix

* Rebase fix

* Modify snapshot calculation for dynamic epoch (#997)

* Modify snapshot calculation for dynamic epoch

* Comments fix

* Comments fix

* Move tests to new file

* Comments fix

* Rebase fix

* Comments fix

* Comments fix

ferranbt committed:
Feature - State Sync manager (#1067)

* initial work

* Rework on event tracker and tests

* More stuff

* Migrate last test

* More changes

* Remove unused test

* Use correct core-contracts version

* Remove logs

* Use eventTracker from tracker package, code cleanup, bug fixes

* Code cleanup

* Code cleanup

* Comments fix

* Dummy state sync manager when bridge is not enabled

* Comments fix

* Comments fix

* Rebase fix

Co-authored-by: Goran Rojovic <goran.rojovic@ethernal.tech>

Commit:
Core contracts to dev-audit (#1095)

* Core contracts to dev-audit
* Fix event name

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/0xPolygon/polygon-edge?pullRequest=1004) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/0xPolygon/polygon-edge?pullRequest=1004) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/0xPolygon/polygon-edge?pullRequest=1004) it.</sub>

Stefan-Ethernal committed:
Merge branch 'develop' into feature/v3-parity

Stefan-Ethernal committed:
Implement `Stringer` interface in `AccountSet` (#1103)

* Implement stringer in AccountSet

* Use strings.Join instead of defining buffer explicitly

begmaroman committed:
Merge remote-tracking branch 'origin/feature/v3-parity' into feature/EVM-299

# Conflicts:
#	.golangci.yml
#	command/loadbot/generator/contract_txn_generator.go
#	command/loadbot/generator/deploy_generator.go
#	command/loadbot/generator/transfer_generator.go
#	command/loadbot/helper.go
#	consensus/polybft/fsm.go
#	consensus/polybft/state_transaction.go
#	jsonrpc/debug_endpoint.go

begmaroman committed:
Minor fixes

Stefan-Ethernal committed:
Add disclaimer message when running polybft consensus protocol (#1107)

* Add disclaimer message when running polybft consensus protocol

* Print disclaimer message to logger and CLI outputter

Stefan-Ethernal committed:
Register validator e2e test (#1084)

* Initial approach

* Rely on system state in TestE2E_Consensus_RegisterValidator

* Add stake and balance flags to register validator cmd

* Remove debug log

* Add appropriate assertions

* Revert scientific notation changes

* Add comments

* Remove grpc flag from register validator cmd

* Delegation and undelegation e2e test (#1086)

* Initial approach

* Delegation logic

* Minor changes to unstake and withdraw commands

* Withdrawal and undelegate in test

* Change unstake cmd description

* Query for current block number before outputting delegator info

Stefan-Ethernal committed:
Handled go routines in event tracker (#1080)

* Manage goroutines in EventTracker

* Stop StateSyncRelayer

* Remove closeCh from polybft and use context instead

* Create named logger in the event tracker itself

* Update consensus/polybft/polybft_test.go assertion message

Co-authored-by: Victor Castell <victor@polygon.technology>

* Remove context field from consensusRuntime

* Isolate changes (revert changes in polybft to rely on channel)

* Add function comments

* Fix test

Co-authored-by: Victor Castell <victor@polygon.technology>

vcastellm commented on types/rlp_unmarshal.go:
```diff
@@ -307,79 +307,156 @@ func (t *Transaction) UnmarshalRLPFrom(p *fastrlp.Parser, v *fastrlp.Value) erro
 
 	if len(elems) < 9 {
-		return fmt.Errorf("incorrect number of elements to decode transaction, expected 9 but found %d", len(elems))
+		return fmt.Errorf("incorrect number of elements to decode transaction, expected at leaset 9 but found %d", len(elems))
```
```suggestion
		return fmt.Errorf("incorrect number of elements to decode transaction, expected at least 9 but found %d", len(elems))
```

begmaroman committed:
Update types/rlp_unmarshal.go

Co-authored-by: Victor Castell <victor@polygon.technology>

begmaroman committed:
Merge remote-tracking branch 'origin/feature/v3-parity' into feature/EVM-299

# Conflicts:
#	command/e2e/register_validator.go

begmaroman committed:
Sync
